Parallel moorings

fsalmon001 opened the pull request:
So far, two options to write a mooring file:
- One file is written sequentially
- One file is written per processor, so N files are written in parallel.

In the proposed pull request, one file is written in parallel using the parallel library netCDF (you need to compile netCDF with the parallel option). NeXtSIM can still be compiled without netcdf parallel but this option could not be used.

The netCDF library can only write rectangular grids efficiently, so there is first a rectangle decomposition of the domain.

Then each processor writes a part of the grid corresponding to one rectangle or a set of rectangles.

I tested it only with a regular grid because I have no file for an arbitrary grid. Please check with an arbitrary grid if you use it, there could be issues in this case.

fsalmon001 committed:
Parallel moorings

tdcwilliams commented:
Thanks @fsalmon001.
I am having trouble compiling the netcdf libraries though.
Which versions of netcdf-c, netcdf-cxx and netcdf-fortran are you using, and where are you downloading them from?

I can compile netcdf-c (latest version, 4.9.3) but netcdf-cxx fails to compile

tdcwilliams commented:
Ok sure thanks @fsalmon001. I'll try to compile netdcf-cxx then. Is version
4.3.1 ok and what flags did you use for configure for that one?

On Tuesday, 21 October 2025, fsalmon001 ***@***.***> wrote:

> *fsalmon001* left a comment (nansencenter/nextsim#682)
> <https://github.com/nansencenter/nextsim/pull/682#issuecomment-3425056353>
>
> In gridouput no, but there are netcdf functions elsewhere, in other
> functions. I did not have a look on that @tdcwilliams
> <https://github.com/tdcwilliams>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nansencenter/nextsim/pull/682#issuecomment-3425056353>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATVYQOM4REL5QXNT5YP5KD3YXLOBAVCNFSM6AAAAACJU5UA42VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMRVGA2TMMZVGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tdcwilliams commented:
Hi @fsalmon001,
I got it to compile and run in my container. For large_arctic_10km.msh and 10 days it was 1 minute faster.

I still need to try an irregular output grid though.

Maybe I'll try large_arctic_5km as well to see the difference there as well.

tdcwilliams committed:
add help for moorings.parallel_output

tdcwilliams committed:
moorings.parallel_output only a valid option if NETCDF_PARALLEL is
defined

tdcwilliams commented:
Hi again @fsalmon001 
I forgot to add `moorings.parallel_output=true` so the run was not using parallel writing. When I set it to true, the model hangs when it it is first time to write to the netcdf file.

This happens both when I use a container and when I compile with intel compilers.
Did this ever happen to you at all?

fsalmon001 commented:
Hi @tdcwilliams, it is with the same version of netcdf as me? I do not understand why it could not run with a null pointer. It is at the first iteration of the loop or after some iterations? What are your options for moorings? Maybe I can test here.

tdcwilliams committed:
fix hang when calling nc_put_vara_float with nullptr

tdcwilliams committed:
Revert "fix hang when calling nc_put_vara_float with nullptr"

This reverts commit 606ced65953d51fda3150f524da39300eee8d5c4.

Only works for 2 cpus. Similarly commenting the nullptr write only works
for 2 cpus.

fsalmon001 commented:
I also asked to some generative AIs. Have you tried like in your commit to initialize `size_t count[3] = {0, 0, 0};` and to use a buffer instead of nullptr for dummy data:
```

                 else // Dummy
                {   

                    std::vector<float> tmp(0);

                    nc_put_vara_float(ncid, data, start, count, &tmp[0]);

                }
```


Sorry I have nothing more @tdcwilliams

tdcwilliams committed:
Merge branch 'parallel_moorings' of github.com:nansencenter/nextsim into parallel_moorings

tdcwilliams commented:
Hi @fsalmon001 
thank you.

I've put the files here:
https://github.com/nansencenter/nextsim-env/tree/apptainer-netcdf-parallel/apptainer

On your personal computer install apptainer (`sudo apt-get install apptainer` for ubuntu) and then do `sudo apptainer build nextsim.sif nextsim.def`

The nextsim.def file has the compilation formulae that  I used for the netcdf libraries.

Incidentally, on our HPC nextsim runs 2x faster inside apptainer than with intel compilers, so I think it would be worth getting it to work for its own sake, and not just for portability,

Once you've built the image files and compiled the model, you may need some more help to run the model. On our HPC, I source this file: https://github.com/nansencenter/nextsim-env/blob/apptainer-netcdf-parallel/machines/tim/fram/env/pynextsim.apptainer.src

which mounts some directories inside the container with forcing data etc and sets some variables inside it.

It works with `moorings.parallel_output=false`, so try that option out first while you're working out running with the container.

tdcwilliams commented:
Hi @fsalmon001 
I am still getting hanging inside this container with `moorings.parallel_output=true` - on our new HPC and also on my laptop.
Maybe I am making some mistake with running it - can you give some more details of how you ran it in your container?
Was it on an HPC or a laptop (if so, what sort)?

fsalmon001 commented:
Hi @tdcwilliams,

I am on my laptop, with ubuntu 22.04.3.

I just go into my container, then `apptainer shell nextsim.sif` and then in the shell I compiled nextsim (the branch moorings_parallel) and I run my script, which includes the different exports of NEXTSIMDIR, NEXTSIM_MESH_DIR, etc, and
`mpirun -np 16 $NEXTSIMDIR/model/bin/nextsim.exec --config-files=file.cfg`

fsalmon001 committed:
Parallel moorings for irregular grids

fsalmon001 commented:
In addition @tdcwilliams, I think I have found a way to deal with the irregular grids. There are more parallel exchanges, so this should be less time efficient, but in terms of memory, it should be similar to the regular grids. It works here.

Moreover, I had an issue with hanging runs. This stemmed from MPI exchanges, so I modified them with non-blocking exchanges. Then, could you try again with this commit using both the regular and irregular grids please? Hopefully this could solve your problem.

tdcwilliams commented:
Hi @fsalmon001 
thanks - that sounds like good progress.

I tried the latest code but it still hangs.

Could you try one more thing with the container please? `apptainer shell --cleanenv --bind path/to/nextsim:/nextsim nextsim.sif` and see if it still works?

tdcwilliams commented:
Thanks @fsalmon001.

einola committed:
Merge remote-tracking branch 'origin/develop' into parallel_moorings

einola committed:
Add variables to help with finding libraries

Adds variable definitions in the Docker and model/Makefile files, so
that the compiler can find netcdf_par.h (and MMG libraries also).

einola committed:
Only add NETCDF_PARALLEL_DIR to CXXFLAGS if the first exists

Adding and empty -I to CXXFLAGS is not a problem, but it's cleaner like
this.

einola committed:
Remove NETCDF_PARALLEL_DIR

It's not needed; we should just point NETCDF_DIR to the parallel library
instead.

einola committed:
Fix an astray \

Didn't test this well enough before pushing last time.

einola committed:
Add timers to the output code

This is to see better how much time it takes to output the moorings.

einola committed:
Merge functions to stay DRY

Merge sequential_initNetCDF and void parallel_initNetCDF into
initNetCDF. It wasn't necessary to have sequential_initNetCDF and void
parallel_initNetCDF, and they contained almost entirely the same
functionality.

einola committed:
Add compression of netCDF output

We've been compressing the files after the fact, but it's better to do
it a t runtime. This should belong to a separate PR, but since we're
changing so much of the moorings code now, I thought I'd slip it in
here.

einola committed:
Merge remote-tracking branch 'origin/develop' into parallel_moorings

tdcwilliams committed:
fix issue with Makefile - should never have -isystem /usr/include in CXXFLAGS

tdcwilliams committed:
Merge branch 'parallel_moorings' of github.com:nansencenter/nextsim into parallel_moorings

tdcwilliams committed:
Merge branch 'parallel_moorings' of github.com:nansencenter/nextsim into parallel_moorings

tdcwilliams committed:
help for option moorings.parallel_output

tdcwilliams committed:
Merge branch 'develop' into parallel_moorings
